README.md
# Krates HAProxy

[![Build Status](https://travis-ci.org/appstersio/krates-haproxy.svg?branch=master)](https://travis-ci.org/appstersio/krates-haproxy)

HAProxy image that does load-balancing between configured addresses. HAProxy reconfigures itself automatically when backend dns address change.

## Configuration

| Environment Variable | Default | Description |
|----------------------|---------|-------------|
| MODE                 | http | Mode of load balancing. Possible values: http, tcp |
| FRONTEND_PORT        | 80      | Port to listen. If `SSL_CERT` is set, will also bind to port 443. |
| BACKENDS             |         | Comma separated list of backends to use, example: app1.foo.com:8080,app2.bar.com:8181 |
| BALANCE              | roundrobin | Load balancing algorithm to use. Possible values include: roundrobin, static-rr, source, leastconn. |
| MAXCONN              | 4096 | Sets the maximum per-process number of concurrent connections. |
| OPTION              | redispatch, forwardfor | Extra options (comma separated list). |
| TIMEOUT              | connect 5000, client 50000, server 50000 | Connect options. |
| POLLING_INTERVAL     | 10 | How often backend addresses are polled (resolved). |
| VIRTUAL_HOSTS        |        | Comma separated list of virtual_host=backend mappings, for example app1=www.domain.com,app2=www.bar.com |
| SSL_CERT             |        | Ssl cert, a pem file with private key followed by public certificate, '\n'(two chars) as the line separator. |


## License

Krates is licensed under the MIT License. See [LICENSE](LICENSE.txt) for full license text.

Kontena software is open source, and you can use it for any purpose, personal or commercial. Kontena is licensed under the Apache License, Version 2.0. See [LICENSE](LICENSE) for full license text.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2020-03-20
### Added
- Docker: Make `/var/lib/acme` a volume by default.
- HAProxy: Use default binary and config file locations as recommended.
- HAProxy: Switch to run in `master-worker` mode and add support for graceful server configuration reload via 'SIGUSR2' signal.
- Makefile: Setup end-2-end developer workflow and enable 'frp' tunnel to validate integration with LetsEncrypt.
- Makefile: Cache locally '/var/lib/acme' and '/etc/ssl/private' folders in './tmp' for introspection and troubleshooting.
- ACME: Switch to use LEGO project to obtain and renew LetsEncrypt certificates for the instance + support for LetsEncrypt staging endpoint in 'debug' mode.

### Changed
- Celluloid: Switch to use newer `supervise` syntax for our supervisors in `bin/haproxy` (celluloud 0.17.4).
- AcmeWorker: Fix issue with `FileUtils.copy` and use `cp` utility instead (via system call).
- HAProxy: Switch to use the most recent released version of the software (`v2.1.3` ).
- HAProxy: Enable 'debug' mode in the logs to improve diagnostics and troubleshooting.
- HAProxy: Get rid of the configuration warnings in the server logs.
- Exclude 'tmp' folder from being tracked by Git.
- HAProxy: Include an actual copy of 'haproxy.cfg' for reference purposes.

### Removed
- ACME: All interactions with 'acmetool' since it has been lagging behind with ACME v2 support.


spec/spec_helper.rb
require_relative '../lib/kontena_haproxy'
Celluloid.logger = nil
Celluloid.shutdown

require 'pry'

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause this
# file to always be loaded, without a need to explicitly require it in any files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    # be_bigger_than(2).and_smaller_than(4).description
    #   # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #   # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Limits the available syntax to the non-monkey patched syntax that is recommended.
  # For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/kontena/backend_resolver_spec.rb
require_relative '../spec_helper'

describe Kontena::BackendResolver do

  before(:each) { Celluloid.boot }
  after(:each) { Celluloid.shutdown }

  let(:subject) {
    described_class.new(interval: 5, backends: ['mysql:3306'])
  }

  describe '.new' do
    it 'sets interval' do
      resolver = described_class.new(interval: 10)
      expect(resolver.interval).to eq(10)
    end

    it 'sets backends' do
      backends = ['mysql:3306']
      resolver = described_class.new(backends: backends)
      expect(resolver.backends).to eq(backends)
    end

    it 'sets dns_resolver' do
      resolver = described_class.new(interval: 10)
      expect(resolver.dns_resolver).to be_instance_of(Celluloid::IO::DNSResolver)
    end
  end

  describe '#start!' do
    it 'calls #resolve_backends' do
      expect(subject.wrapped_object).to receive(:resolve_backends)
      allow(subject.wrapped_object).to receive(:resolve_dns)
        .with('mysql').and_return(['10.81.1.5'])
      subject.async.start!
      sleep 0.01
    end
  end

  describe '#resolve_backends' do
    it 'publishes resolved ips' do
      expect(subject.wrapped_object).to receive(:publish).with('backends_resolved', {'mysql' => [{ip: '10.81.1.5', port: '3306'}]})
      allow(subject.wrapped_object).to receive(:resolve_dns)
        .with('mysql').and_return(['10.81.1.5'])
      subject.resolve_backends
    end
  end

  describe '#resolve_dns' do
    it 'calls Celluloid::IO::DnsResolver' do
      expect(subject.dns_resolver).to receive(:resolve).with('www.google.com').once
      subject.resolve_dns('www.google.com')
    end

    it 'returns array of ips' do
      ips = subject.resolve_dns('www.google.com')
      expect(ips.size).to be > 0
    end

    it 'returns passed value if dns does not resolve' do
      ips = subject.resolve_dns('10.1.1.1')
      expect(ips[0].to_s).to eq('10.1.1.1')
    end
  end
end


lib/kontena/acme_worker.rb
module Kontena
  class AcmeWorker
    include Celluloid
    include Celluloid::Logger
    include Celluloid::Notifications

    attr_reader :email, :domains, :debug

    ACME_CMD = "lego"
    ACME_CERT_DIR = "/var/lib/acme/"
    ACME_LOGS_DIR = ACME_CERT_DIR + "logs"
    ACME_LOG_FILE = "#{ACME_LOGS_DIR}/console.log".freeze

    # @param [String] email
    # @param [Array<String>] domains
    # @param [Boolean] debug
    # @param [Boolean] autostart
    def initialize(email, domains, debug = false, autostart = true)
      # Local copy of the incoming parameters
      @domains, @email, @debug = domains, email, debug
      # Ensure logs folder exists before using it
      FileUtils.mkdir_p(ACME_LOGS_DIR) unless Dir.exist?(ACME_LOGS_DIR)
      # Switch to use staging endpoint to avoid hitting LE's prod rate limits
      ACME_CMD << " -s https://acme-staging-v02.api.letsencrypt.org/directory" if debug
      # Kick-start the main routine in case of auto-start
      async.start! if autostart
    end

    def start!
      defer {
        domains.each{|d|
          want_domain(d)
          copy_domain_cert(d)
        }
        publish 'haproxy:config_updated'
        loop do
          sleep (60*60*24*7) unless debug # week
          sleep (60*2) if debug # 2 minutes
          domains.each{|d|
            reconcile_domain(d)
            copy_domain_cert(d)
          }
          publish 'haproxy:config_updated'
        end
      }
    end

    # @param [String] domain
    # @return [Boolean]
    def want_domain(domain)
      retries = 0
      begin
        # lego -a -m <email> -d <domain> --http --http.port 127.0.0.1:402 --path <state> run
        success = system("#{ACME_CMD} -a -m #{email} -d #{domain} --http --http.port 127.0.0.1:402 --path #{ACME_CERT_DIR} --pem run >> #{ACME_LOG_FILE}")
        if success
          info "Fetched cert for domain #{domain}, expecting 'lego' to finish write operations by that time"
        else
          retries += 1
          raise "failed to fetch cert for domain #{domain}"
        end
      rescue => exc
        info exc.message
        # In case of an exception, display acme logs as well
        info File.read ACME_LOG_FILE
        wait = 10 * retries
        info "Retrying certificate fetch for '#{domain}' in #{wait} seconds"
        sleep wait
        retry
      end
    end

    # @param [String] domain
    def copy_domain_cert(domain)
      file_path = ACME_CERT_DIR + "certificates/#{domain}.pem"
      # Each pem file at the destination must have timestamp suffix
      ssl_path = "/etc/ssl/private/#{domain}.pem"
      if File.exist?(file_path)
        info "Copying #{domain} certificate to '#{ssl_path}'"
        system("cp -fv #{file_path} #{ssl_path} >> #{ACME_LOG_FILE}")
      elsif File.exist?(ssl_path)
        system("rm -fv #{ssl_path} >> #{ACME_LOG_FILE}")
        info "Removing certificate from #{domain} at '#{ssl_path}'"
      end
    end

    # @return [Boolean]
    def reconcile_domain(domain)
      retries = 0
      begin
        # lego -m <email> -d <domain> --http --path /var/lib/acme --pem renew
        success = system("#{ACME_CMD} -m #{email} -d #{domain} --http --http.port 127.0.0.1:402 --path #{ACME_CERT_DIR} --pem renew #{'--days 999' if debug} >> #{ACME_LOG_FILE}")
        if success
          info "Renewed cert for domain #{domain}, expecting 'lego' to finish write operations by that time"
        else
          retries += 1
          raise "Failed to renew cert for domain #{domain}"
        end
      rescue => exc
        info exc.message
        # In case of an exception, display acme logs as well
        info File.read ACME_LOG_FILE
        wait = 10 * retries
        info "Retrying certificate renewal for '#{domain}' in #{wait} seconds"
        sleep wait
        retry
      end
    end
  end
end


lib/kontena/haproxy_spawner.rb

module Kontena
  class HaproxySpawner
    include Celluloid
    include Celluloid::Logger
    include Celluloid::Notifications

    ##
    # @param [String] config_file
    def initialize(haproxy_bin = '/usr/local/sbin/haproxy', config_file = '/usr/local/etc/haproxy/haproxy.cfg')
      @current_pid = nil
      @haproxy_cmd = [haproxy_bin, '-f', config_file, '-db', '-d']
      subscribe 'haproxy:config_updated', :update_haproxy
      info '~~ Starting Krates HAProxy ~~'
    end

    def update_haproxy(*args)
      if current_pid
        reload_haproxy
      else
        start_haproxy
      end
    end

    def start_haproxy
      info "Starting HAProxy process ~> '#{@haproxy_cmd.join(' ')}'"
      @current_pid = Process.spawn(@haproxy_cmd.join(' '))
    end

    def reload_haproxy
      info "Requesting graceful HAProxy configuration reload via SIGUSR2 (PID: #{@current_pid})"
      Process.kill("USR2", @current_pid)
      info "Graceful HAProxy configuration reload is done (PID: #{@current_pid})"
    end

    private

    def current_pid
      @current_pid
    end
  end
end


lib/kontena/backend_resolver.rb

module Kontena
  class BackendResolver
    include Celluloid
    include Celluloid::Logger
    include Celluloid::Notifications

    attr_reader :interval, :backends, :dns_resolver

    def initialize(opts)
      @interval = opts[:interval].to_i
      @backends = opts[:backends]
      @dns_resolver = Celluloid::IO::DNSResolver.new
      async.start! if opts[:autostart]
    end

    def start!
      info "Starting to resolve ip's for backends: #{self.backends}"
      info "with interval of #{self.interval} seconds"
      defer {
        loop do
          resolve_backends
          sleep self.interval
        end
      }
    end

    def resolve_backends
      resolved_backends = {}
      self.backends.each do |backend|
        service, port = backend.split(':')
        resolved_ips = resolve_dns(service)
        debug resolved_ips
        if resolved_ips
          resolved_ips = resolved_ips.sort_by{ |ip| ip.to_s }
          resolved_backends[service] = resolved_ips.map{|r|
            { ip: r.to_s, port: port }
          }
        end
      end
      publish 'backends_resolved', resolved_backends
    rescue => exc
      error "Failed to resolve backends: #{exc.message}"
      error "Failed to resolve backends: #{exc.backtrace.join("\n")}"
    end

    ##
    # @param [String] dns_name
    # @return [Array<Resolv::IPv4>]
    def resolve_dns(dns_name)
      resolved = dns_resolver.resolve(dns_name)
      if resolved.is_a?(Resolv::IPv4)
        [resolved]
      elsif resolved.nil?
        [dns_name]
      else
        resolved
      end
    end
  end
end


lib/kontena/haproxy_config_writer.rb
module Kontena
  class HaproxyConfigWriter
    include Celluloid
    include Celluloid::Logger
    include Celluloid::Notifications

    attr_accessor :config_file

    ##
    # @param [String] config_file
    def initialize(config_file = '/usr/local/etc/haproxy/haproxy.cfg')
      self.config_file = config_file
      @old_config = ''
      subscribe 'haproxy:config_generated', :update_config
    end

    ##
    # @param [String] topic
    # @param [String] config
    def update_config(topic, config)
      if @old_config != config
        info config
        write_config(config)
        @old_config = config
        publish 'haproxy:config_updated'
      end
    end

    ##
    # @param [String] config
    def write_config(config)
      File.write(config_file, config.to_s)
    end
  end
end


lib/kontena/haproxy_config_generator.rb
module Kontena
  class HaproxyConfigGenerator
    include Celluloid
    include Celluloid::Logger
    include Celluloid::Notifications

    attr_reader :options, :config, :vhosts

    def initialize(opts)
      @options = opts
      @config = {}
      create_vhosts
      create_default_config
      subscribe 'backends_resolved', :update_config
    end

    def create_default_config
      config['global'] = [
        "maxconn %s" % options[:maxconn],
        "pidfile /var/run/haproxy.pid",
        "user haproxy",
        "group haproxy",
        "master-worker",
        "stats socket /var/run/haproxy.stats level admin",
        "tune.ssl.default-dh-param 2048"
      ]
      config['defaults'] = [
        "mode %s" % options[:mode]
      ]
      if options[:option]
        options[:option].each do |option|
          config['defaults'] << 'option %s' % option.strip
        end
      end
      if options[:timeout]
        options[:timeout].each do |timeout|
          config['defaults'] << 'timeout %s' % timeout.strip
        end
      end
    end

    ##
    # @param [String] topic
    # @param [Hash] backends
    def update_config(topic, backends)
      create_frontend
      create_backend(backends)
      publish('haproxy:config_generated', config_to_string)
    end

    def create_frontend
      frontend = []
      frontend << 'bind 0.0.0.0:%s' % options[:frontend_port]
      if has_certs?
        frontend << 'bind 0.0.0.0:443 ssl crt /etc/ssl/private/'
        frontend << 'http-request add-header X-Forwarded-Proto https'
        frontend << 'redirect scheme https code 301 if !{ ssl_fc } !{ url_beg /.well-known/acme-challenge/ }'
      end

      frontend << 'acl acme url_beg /.well-known/acme-challenge/'
      frontend << 'use_backend acme if acme'

      if options[:virtual_hosts] != ''
        vhosts.each do |service, domain|
          id = service.downcase.sub('-', '_')
          frontend << 'acl host_%s hdr(host) -i %s' % [id, domain]
          frontend << 'use_backend %s_cluster if host_%s' % [id, id]
        end
      else
        frontend << 'default_backend default_backend'
      end
      config['frontend default_frontend'] = frontend
    end

    ##
    # @param [Hash] backend_services
    def create_backend(backend_services)
      if options[:virtual_hosts] != ''
        create_virtual_host_backends(backend_services)
      else
        create_default_backend(backend_services)
      end

      config['backend acme'] = [
        'server acmetool 127.0.0.1:402'
      ]
    end

    ##
    # @param [Hash] backend_services
    def create_virtual_host_backends(backend_services)
      vhosts.each do |service, domain|
        if backend_services[service]
          backend = []
          backend << 'balance %s' % options[:balance]
          service_name = service.gsub('.', '-')
          i = 1
          backend_services[service].each do |host|
            backend << 'server %s %s:%s' % ["#{service_name}-#{i}", host[:ip], host[:port]]
            i += 1
          end
          id = service.downcase.sub('-', '_')
          config["backend %s_cluster" % id] = backend
        end
      end
    end

    ##
    # @param [Hash] backend_services
    def create_default_backend(backend_services)
      backend = []
      backend << 'balance %s' % options[:balance]
      backend_services.each do |service, hosts|
        service_name = service.gsub('.', '-')
        i = 1
        hosts.each do |host|
          backend << 'server %s %s:%s' % ["#{service_name}-#{i}", host[:ip], host[:port]]
          i += 1
        end
      end
      config['backend default_backend'] = backend
    end

    def create_vhosts
      @vhosts = {}
      options[:virtual_hosts].split(',').each do |host|
        service, domain = host.split('=', 2)
        if service && domain
          @vhosts[service.strip] = domain.strip
        end
      end
    end

    ##
    # @return [String]
    def config_to_string
      conf = ''
      config.each do |key, values|
        conf << "#{key}\n"
        values.each do |v|
          conf << "  #{v}\n"
        end
      end
      conf
    end

    # @return [Boolean]
    def has_certs?
      Dir.glob("/etc/ssl/private/*").size > 0
    end
  end
end
